kubeadm implement preflight checks

dmmcquay opened the pull request:
Checks that user running kubeamd init and join is root and will only execute
command if user is root. Moved away from using kubectl error handling to
having kubeadm handle its own errors. This should allow kubeadm to have
more meaningful errors, exit codes, and logging for specific kubeadm use
cases.

fixes #33908

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34341)

<!-- Reviewable:end -->

dmmcquay committed:
kubeadm implement preflight checks

Includes checks for verifying services exist and are enabled, ports are
open, directories do not exist or are empty, and required binaries are
in the path.

Checks that user running kubeamd init and join is root and will only execute
command if user is root. Moved away from using kubectl error handling to
having kubeadm handle its own errors. This should allow kubeadm to have
more meaningful errors, exit codes, and logging for specific kubeadm use
cases.

dgoodwin committed:
Add kubeadm preflight check framework.

Includes checks for verifying services exist and are enabled, ports are
open, directories do not exist or are empty, and required binaries are
in the path.

pires commented:
@dmmcquay please rebase.

pires commented:
Also, if merged, fixes #33928.

luxas commented on cmd/kubeadm/app/util/user.go:
```diff
@@ -0,0 +1,15 @@
+)
+
+func IsRoot() (*user.User, bool) {
+	u, err := user.Current()
```
No please don't do this.
It's a total PR blocker for me, since it will bring in a lot of C code which will break the static link and some OSes like GCI.

For more information, see: #26028. We should probably document somewhere that use of `os/user` is a big no-no if not _really_ required.

Use `os.Getuid()` instead.

luxas commented on cmd/kubeadm/app/kubeadm.go:
```diff
@@ -17,21 +17,42 @@ limitations under the License.
-`)
+// TODO(phase2) use componentconfig
+// we need some params for testing etc, let's keep these hidden for now
+func getEnvParams() map[string]string {
```
This is probably a rebase conflict, since `GetEnvParams()` is now in `cmd/kubeadm/app/api/env.go`

luxas replied:
ping. remove this

luxas commented on cmd/kubeadm/app/cmd/cmd.go:
```diff
@@ -62,8 +62,6 @@ func NewKubeadmCommand(f *cmdutil.Factory, in io.Reader, out, err io.Writer) *co
 
 		`),
 	}
-	// TODO(phase2+) figure out how to avoid running as root
```
This should be left as-is, since we still should figure out how to remove this dependency

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -48,8 +47,8 @@ func NewCmdInit(out io.Writer) *cobra.Command {
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunInit(out, cmd, args, cfg)
-			cmdutil.CheckErr(err)
+			err := RunInit(out, cmd, args, s, advertiseAddrs)
```
Seems like a rebase conflict here

luxas commented on cmd/kubeadm/app/util/user.go:
```diff
@@ -0,0 +1,15 @@
+	sudoUser = "0"
+)
+
+func IsRoot() (*user.User, bool) {
```
I'd prefer to move this function into a separate package: `cmd/kubeadm/app/preflight`

pires replied:
@luxas can it be moved to said package in bigger _pre-flight_ PR, which I believe you and @dgoodwin are working on?

luxas commented on cmd/kubeadm/app/util/error.go:
```diff
@@ -0,0 +1,114 @@
+
+const (
+	DefaultErrorExitCode = 1
+	EX_USAGE             = 64
```
Where do these come from?
Do we need everything?

dmmcquay replied:
Yes they are from sysexits.h. I also agree that we don't need all of them, and even perhaps changing their names to be more clear what they are would be a good idea.

luxas commented on cmd/kubeadm/app/util/error.go:
```diff
@@ -0,0 +1,114 @@
+	kubeadm: and make sure to mention @kubernetes/sig-cluster-lifecycle. Thank you!
+`)
+
+type NotRootError struct {
```
Do you plan adding structs for every possible error message?
Seems like that's going to be a maintenance heavy thing to do.

I think I'm in favor for plain Go-style error handling

dmmcquay replied:
I looked at your open PR #34535 and I think we could move towards a consensus between the work there and some of the ideas I have in this PR. 

I am not suggesting that we do a struct per error approach on all errors.  There are going to be some errors that we want to handle a specific way and others that are more general. In my opinion, the only way to be able to handle diverse errors in different ways (different exit codes, etc.) is not to inspect the contents of the error (something like strings.Contains()), but be able to detect what type of error they are. Which is why structs that implement the error interface would work great in this case. (See http://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully)

I personally don't like that you pass around []string, I think they easily could be a []error.  Also, in your PR, there are more than one place where errors are handled and also more than one place where you program may exit.  I think it is best practice to have one place that handles errors and determines how the program should exit (errors codes, logging messages, etc). 

If we moved towards having an []error rather than []string, one could iterate over the errors and handle each accordingly.  If some only require some logging, they just log, while if others require more action, they can do so. 

As for glog, I am not personally attached to it.  Although, logs that are more machine readable are helpful, something the default log packages doesn't handle great.  

I think in the long run, getting this right is not so much for humans as it is for automating this process.  Meaningful exit codes allow things to be scripted and the returning errors codes have a lot of value in allowing said system to be designed.

luxas reviewed (changes requested):
The `os/user` thing is a blocker for me.

Overall, this seems like much code for a very little check, but if you can convince us that this new glog-like tool in error.go is better than anything else, we might go for it.

I still want @errordeveloper and @mikedanese take a look and say what they're thinking

luxas commented:
@k8s-bot ok to test

dmmcquay commented:
@luxas using glog and creating meaningful error structs  is nothing new.  This is what kubeadm used to handle its error messaging before (passing them off to kubectl).  I don't think there is a need to convince that this is how things should be done since thats how it previously was done.  I just think kubeadm should be able to handle its specific use case errors that other parts of the project won't have. 

Working on fixing requested changes.

dgoodwin commented:
@dmmcquay my WIP is here: #34535

Looks like we should sync up, I skipped the root check as I knew there were already PR's implementing this.

pires commented:
@dgoodwin have you CNCF CLA?

@luxas anything else pending from your changes request?

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -112,12 +113,30 @@ func NewCmdInit(out io.Writer) *cobra.Command {
 	)
 	cmd.PersistentFlags().MarkDeprecated("external-etcd-keyfile", "this flag will be removed when componentconfig exists")
+	cmd.PersistentFlags().BoolVar(
+		&skipChecks, "skip-checks", false,
```
maybe `skip-preflight` instead. I think that sounds better

pires replied:
`skip-checks` has a clearer intent than `skip-preflight`. As an user, if I didn't know already what you're talking about, I'd ask _what's preflight?_.
I'm ok with `skip-preflight-checks`, though.

luxas commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -57,12 +58,31 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
 	)
 
+	cmd.PersistentFlags().BoolVar(
+		&skipChecks, "skip-checks", false,
```
skip-preflight

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+func (e *PreFlightError) Error() string {
+	return fmt.Sprintf("PreFlight Error\n  count: %d \n  msg: %s", e.Count, e.Msg)
```
nit: Pre-flight error or just pre-flight error

(in some parts of the k8s code it's best practice to have lowercase error messages, but we don't have to obey that here I think)

luxas commented on cmd/kubeadm/app/preflight/checks.go:
add more inpathchecks:
ip: true, tc: false, nsenter: true, touch: false, iptables: true, mount: true, ebtables: true

for both master and node

luxas commented on cmd/kubeadm/app/preflight/initsystem.go:
```diff
@@ -0,0 +1,83 @@
+limitations under the License.
+*/
+
+package preflight
```
This package should be moved to `pkg/util/initsystem` as @freehan wants to reuse it for other purposes as well: #34494

dmmcquay replied:
@luxas I don't know where you want this. You want to create a kubeadm/pkg/util/initsystem package? If that is the case, I'm not so sure that is a good naming scheme but I could be wrong.

luxas commented on cmd/kubeadm/app/preflight/user.go:
```diff
@@ -0,0 +1,43 @@
+	return fmt.Sprintf("%s: Current User Uid: %d", e.Msg, e.Uid)
+}
+
+func IsRoot() (int, bool) {
```
it's very unexpected to return both `uid, uid != 0` in this function.
Switch to just
```go
return os.Getuid() == 0
```

also much simpler

luxas replied:
I still think that this function should _only_ return `bool`
When the function name is `IsRoot` the return value should be boolean

If the function name would be `GetUIDAndIfRoot` (which would be terrible), the return value could be `(int, bool)`

luxas commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -57,12 +58,31 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
-func RunJoin(out io.Writer, cmd *cobra.Command, args []string, s *kubeadmapi.NodeConfiguration) error {
+func RunJoin(out io.Writer, cmd *cobra.Command, args []string, s *kubeadmapi.NodeConfiguration, skipChecks bool) error {
 	// TODO(phase1+) this we are missing args from the help text, there should be a way to tell cobra about it
+	if u, ok := preflight.IsRoot(); !ok {
```
somewhere we should think about how to minimize duplication between `RunInit` and `RunJoin`

not required to do it now, but anyway

luxas reviewed (changes requested):
I had some more comments, please fix them and I'll merge

luxas commented:
@dgoodwin please explicitely say you're fine with your commit being contributed in this PR

luxas commented:
Also, the flag `skip-preflight` should be added to `hack/verify-flags/known-flags.txt`

errordeveloper commented:
Ok, the title of this PR doesn't seem to reflect the changes it introduces. It looks like we need to split this up into 3 or more PRs.

dgoodwin commented:
I am good with my commit going in via this PR.

dmmcquay commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -112,12 +113,30 @@ func NewCmdInit(out io.Writer) *cobra.Command {
+func RunInit(out io.Writer, cmd *cobra.Command, args []string, cfg *kubeadmapi.MasterConfiguration, skipChecks bool) error {
+	if u, ok := preflight.IsRoot(); !ok {
+		return &preflight.NotRootError{"<cmd/init> Requires root priviledges", u}
+	}
```
I think we could do that, but in my opinion it's nice to have it return EX_NOPERM 77 for now.

dgoodwin replied:
Is it not possible for it to be in the normal list of checks, and still return it's own code?

pires replied:
@dmmcquay I can support both yours and @dgoodwin opinions. Make sure you agree on one.

pires commented on cmd/kubeadm/app/util/error.go:
```diff
@@ -0,0 +1,95 @@
+		handleErr(err.Error(), PreFlight)
+	default:
+		fmt.Printf(AlphaWarningOnExit)
+		handleErr(err.Error(), DefaultErrorExitCode)
```
@dgoodwin I'm genuinely sure this is going to be needed as even `kubeadm` is a simple tool to use, admins will automate its usage. If we don't return error codes, how will they see it? ;)

dgoodwin replied:
To be clear I mean specific error codes, per pre-flight check. Obviously we fail an exit with an error code on a preflight check that is classified as an error, and the admin sees the output indicating what the problem was. I'm just not sure they're going to write automation around the exact precise error code to take action, i.e. we got a 77, lets re-run as root. Instead I would expect they'd just fix the environment and try again rather than coding that logic into whatever automation they're doing.

pires replied:
Ah, thanks for clarifying. I can support that.

errordeveloper replied:
Totally, we will run out of numbers quite quickly too! :)

pires commented on cmd/kubeadm/app/preflight/checks.go:
Firewalling can be done at the host level or with networking equipment we can't control. However, we need to make sure that whatever connects to this ports has proper error-handling in the case these ports are not _open_ for service.

dgoodwin commented:
My mistake I did not realize there was a new CLA in play, I think I got it signed now. (it was not a clean operation so I am not sure)

pires commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -112,12 +113,30 @@ func NewCmdInit(out io.Writer) *cobra.Command {
+	cmd.PersistentFlags().BoolVar(
+		&skipChecks, "skip-checks", false,
+		"skip checks normally run before modifying the system",
+	)
```
@luxas mentioned this as well and I am not taking this lightly. As an user, I think _pre-flight_ is less clear about the intents of this flag, which is to perform checks. I suggest `skip-preflight-checks`.
That said, all log entries related to pre-flight should include a reference to _pre-flight check {failed,succeeded}_.

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -112,12 +113,30 @@ func NewCmdInit(out io.Writer) *cobra.Command {
 	)
 	cmd.PersistentFlags().MarkDeprecated("external-etcd-keyfile", "this flag will be removed when componentconfig exists")
+	cmd.PersistentFlags().BoolVar(
+		&skipPreFlight, "skip-preflight", false,
```
okay @pires, `skip-preflight-checks` SGTM

luxas commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -0,0 +1,223 @@
+}
+
+func (e *PreFlightError) Error() string {
+	return fmt.Sprintf("preflight error\n  count: %d \n  msg: %s", e.Count, e.Msg)
```
nit: `preflight check error: `

luxas commented on cmd/kubeadm/app/preflight/checks.go:
nit: sort by mandatory and then alphabetically

pires commented:
@luxas i believe every nit was addressed. Here's hoping for your /lgtm.

luxas commented:
@k8s-bot test this

pires commented:
Perhaps I can learn something here. What's the problem with tests?

pires commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -57,12 +58,27 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
 	)
 
+	cmd.PersistentFlags().BoolVar(
+		&skipPreFlight, "skip-preflight-checks", false,
```
New flags must be added to `hack/verify-flags/known-flags.txt`, otherwise build won't pass.

pires commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -0,0 +1,239 @@
+}
+
+func (sc ServiceCheck) Check() (warnings, errors []error) {
+
```
Remove empty line.

pires commented on pkg/util/initsystem/initsystem.go:
```diff
@@ -0,0 +1,83 @@
+)
+
+type InitSystem interface {
+
```
Remove empty line.

dgoodwin committed:
Fix errors and improve output in kubeadm pre-flight checks.

Add skip-preflight-checks to known flags.
Fix bug with preflight checks not returning system is-active as errors.
Fix error handling to use correct function.

dgoodwin commented:
@dmmcquay I made the changes requested by @pires and pushed to my original "preflight" branch. I reset it to your branch, then applied this on top. It adds a third commit unfortunately, ping me on slack if you'd like and we can work out a way to get it back down to 2.

dgoodwin commented on cmd/kubeadm/app/preflight/checks.go:
```diff
@@ -0,0 +1,239 @@
+				sc.service, sc.service))
+	}
+
+	return warnings, nil
```
I knew I was seeing something weird here, I failed to return the errors list we build up.

dgoodwin commented:
Ok I have rebased this on master, fixed a couple issues I introduced in the merge, added the known flag, fixed the whitespace issue, and found/fixed a bug in my commit where we're not returning errors in systemd is-active check. 

@dmmcquay I pushed all of this to a "preflight-2" branch in my repo for your consideration. I wouls suggest taking that branch and force pushing here. I think we're stuck with 3 commits at this point. In the rebase I broke your commit signature so that would need to be re-signed if you wish to do so.

Right now the output looks like:

```
$ kubeadm init master
<cmd/init> Running pre-flight checks
<preflight/checks> WARNING: kubelet service is not enabled, please run 'systemctl enable kubelet.service'

kubeadm: I am an alpha version, my authors welcome your feedback and bug reports
kubeadm: please create issue an using https://github.com/kubernetes/kubernetes/issues/new
kubeadm: and make sure to mention @kubernetes/sig-cluster-lifecycle. Thank you!
<cmd/init> preflight check error
  count: 1 
  msg: 
        kubelet service is not active, please run 'systemctl start kubelet.service'
```

I'm thinking that "I'm an alpha version" is very out of place here and should either be at the start of execution, or not there at all.

Also the "count" and "msg" prefix should both be dropped IMO.

pires commented:
Good job @dgoodwin!

> Also the "count" and "msg" prefix should both be dropped IMO.
I agree but it may happen in a separate PR related to #34559.

luxas commented:
Waiting for @dmmcquay to update his branch before merging

pires commented:
@luxas this is ready to merge.

errordeveloper commented:
Glad we got it rebased! I'll take another pass at the code, if you don't mind...

errordeveloper commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -49,6 +50,7 @@ var (
 func NewCmdInit(out io.Writer) *cobra.Command {
 	cfg := &kubeadmapi.MasterConfiguration{}
 	var cfgPath string
+	var skipPreFlight bool
```
Should be added to config, but I'm happy if you do this in a separate PR.

errordeveloper commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -142,6 +148,17 @@ func NewInit(cfgPath string, cfg *kubeadmapi.MasterConfiguration) (*Init, error)
+			return nil, err
+		}
+	} else {
+		fmt.Println("Skipping pre-flight checks")
```
Missing `<cmd/init>` prefix here...

errordeveloper commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -57,12 +58,27 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
+func RunJoin(out io.Writer, cmd *cobra.Command, args []string, s *kubeadmapi.NodeConfiguration, skipPreFlight bool) error {
 	// TODO(phase1+) this we are missing args from the help text, there should be a way to tell cobra about it
+	if !skipPreFlight {
+		fmt.Println("Running pre-flight checks")
```
Also prefix here...

errordeveloper commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -57,12 +58,27 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
 
+	cmd.PersistentFlags().BoolVar(
+		&skipPreFlight, "skip-preflight-checks", false,
+		"skip preflight checks normally run before modifying the system",
```
I thought the convention was that descriptions start with a capital letter...

errordeveloper commented on cmd/kubeadm/app/cmd/join.go:
```diff
@@ -57,12 +58,27 @@ func NewCmdJoin(out io.Writer) *cobra.Command {
+			return err
+		}
+	} else {
+		fmt.Println("Skipping pre-flight checks")
```
Ditto

pires commented:
@errordeveloper I did take another pass too and tested without `kubelet`.

errordeveloper commented on pkg/util/initsystem/initsystem.go:
```diff
@@ -0,0 +1,80 @@
+
+func (sysd SystemdInitSystem) ServiceExists(service string) bool {
+	args := []string{"status", service}
+	outBytes, _ := exec.Command("systemctl", args...).Output()
```
Just being curios: what would dbus/native code look like for this?

errordeveloper commented on cmd/kubeadm/app/preflight/checks.go:
Also, out out curiosity: This looks good, but we could also consider if conntrack would work for us (don't we have a conntrack wrapper already or depend on it somehow?). Also, netstat could do as well, WDYT?

errordeveloper commented on cmd/kubeadm/app/preflight/checks.go:
Just a heads-up: once #34719 gets merged, there will be constants for default port and config field for .
Also, you don't have etcd and 9898 for disco ports in this list, should we check them? Etcd may be tricky when it comes to external one, it could as well run on the same host, but port check would be critical for default path... To be clear, you don't have to fix this now :)

errordeveloper commented on cmd/kubeadm/app/preflight/checks.go:
We don't care about 8080 on nodes.

errordeveloper reviewed (approved):
Few comments and questions.

errordeveloper commented:
/lgtm

errordeveloper commented:
Considering we won't have time to make a release today, may be we could fix cosmetic bits like the `<cmd/init>` prefix, and port 8080 in node check.

luxas commented on cmd/kubeadm/app/preflight/checks.go:
Yup, I've documented them in https://github.com/kubernetes/kubernetes/issues/26093

pires commented:
@luxas @errordeveloper we keep kicking this down the road and it will just make it harder for the good guys who have worked on this to get it merged when compared to other simpler PR's that are getting merged. That said, can we have this merged?
